### PR TITLE
Adding Class II HLA typing to immuno

### PR DESCRIPTION
**Pull request by Layth17**
- Environment

- [x] Docker image `docker(mgibio/phlat:1.1_withindex)` (note: mem must be >= 16G for bowtie2 to work properly. I ran it with 20G) `bsub -Is -M 20G -G $GROUP -n 1 -R 'select[mem>20000] rusage[mem=20000] span[hosts=1]' -q oncology-interactive -a 'docker(mgibio/phlat:1.1_withindex)' /bin/bash`

- [x] Within docker, bash script `run.b38.sh` is working.

- [x] Create `phlat.wdl`

- [x] Implement `phlat.wdl` within `immuno.wdl`

- Input/Output cleaning

- [x] Extract Alleles -- extract class I from OptiType and class II from PHLAT

-----
testing the steps below will require me to switch pVACtools docker image to laljorani20/pvactools:latest (temporary until pvactools is updated)

- [x] HLA Consensus -- must output a list of *valid* HLAs for class I and II
Relevant sources:
https://github.com/griffithlab/pVACtools/issues/375
https://github.com/griffithlab/pVACtools/issues/528
https://github.com/griffithlab/pVACtools/pull/813 (HLA class II pairing)
For class II, `HLA-` prefix not needed and in fact would cause issues.

- [x] pVACseq and pVACfuse working properly with the input provided

- [x] grab the outputs into final_output

### CODE REVIEW

File: phlat/run.b38.sh
Comment (chrisamiller) [inline phlat/run.b38.sh:149]: In the context of the larger pipelines, (as opposed to a one-off bash script it came from) I don't like that this has a hardcoded human build38 path. (if we tried to run this on b37 data (or mouse data), it would happily do the wrong thing!).  The ideal thing to do here would probably be to use the annotation build info to look up the HLA locus, I think?  That could look something like this:
1) grep the GTF file for the proper gene names (`HLA-A, HLA-B, HLA-C,`, etc) to get coordinates
2) use those coordinates in this command

**Comment by Layth17** [comment]
`phlat` folder should be deleted ultimately. it is only in this repo atm for quick testing purposes.

File: definitions/tools/extract_hla_alleles.wdl
Comment (tmooney) [inline definitions/tools/extract_hla_alleles.wdl:25]: If you want, this can be done without a temporary file by grouping the commands together for the pipe.  Something like:

```bash
(/usr/bin/awk ... ~{file}; grep ... ~{phlat_file} | /usr/bin/awk ...) | /usr/bin/awk ...
```

File: definitions/tools/pvacfuse.wdl
Comment (tmooney) [inline definitions/tools/pvacfuse.wdl:30]: We should switch this back to an official release before merging.

File: definitions/tools/pvacseq.wdl
Comment (tmooney) [inline definitions/tools/pvacseq.wdl:53]: (Same here.)

File: definitions/tools/phlat.wdl
Comment (tmooney) [inline definitions/tools/phlat.wdl:21]: Does the boot disk need to scale this severely with the size of the input data?

File: definitions/tools/pvacseq.wdl
Comment (tmooney) [inline definitions/tools/pvacseq.wdl:122]: What is intended and what is happening instead?

File: definitions/tools/phlat.wdl
Comment (Layth17) [reply definitions/tools/phlat.wdl:21]: I think I was mimicking optitype.wdl when I created phlat.wdl

If there is a better way to do this, I am all for it. Would times 2 be better?

File: definitions/tools/pvacseq.wdl
Comment (Layth17) [reply definitions/tools/pvacseq.wdl:122]: From the looks of the the command that uses glob, it seemed like it intends to iterate through pvacfuse_predictions/ like this:

| PVACfuse_predictions/
  | mhc_i
    *grabs all files*
  | mhc_ii
    *grabs all files*
  | combined
    *grabs all files*

And output a similar structure to the above.

However the the real result of that was like this:

| PVACfuse_predictions/
   | mhc_i
     *grabs all files*

Leading to the output:

PVACfuse_predictions/<content of mhc_i>

And the documentation does state that glob skips folders and their contents. Only thing that is slightly vague to me is the use of the double * and why did it pick mhc_i as opposed to another folder unless that was just purely due to order.

File: definitions/tools/phlat.wdl
Comment (malachig) [reply definitions/tools/phlat.wdl:21]: I'm working on a resource usage script that I hope will help us see how much disk space is actually being used so that we can fine tune some of these resource requests.

File: definitions/tools/pvacfuse.wdl
Comment (malachig) [reply definitions/tools/pvacfuse.wdl:30]: I believe this is blocked on @susannasiebert switching back to pvactools effort and creating a release. But yeah we don't want to be pointing to personal docker images like this.  I have been meaning to search the whole repo and see if we are doing it anywhere else.

**Review by malachig — CHANGES_REQUESTED** [review]
As mentioned by Tom above, I think we want to switch to a docker image for an official pvactools release before we consider this final.

**Comment by malachig** [comment]
I have tested this PR on real patient data and it seems to be working nicely:
- I got a successful immuno.wdl run.
- I manually examined the phlat HLA typing results and compared class I calls to optitype calls (100% agreement)
- I examined the organization of class I and II HLA typing results into final folders with improved file names.  This looks good to me
- I confirmed that class I pvacseq and pvacfuse results are using the class I HLA types predicted
- I confirmed that class II pvacseq and pvacfuse results are using the class II HLA types predicted
- I loaded both the class I and class II neoantigen results in pvacview and manually examined them

I think other than the issue with using a personal docker image for pvactools this is ready to merge...  We should ask @susannasiebert what she thinks about a timeline for a release that would include the pvactools improvements needed here.  I believe the main thing was the code that produces classII allele pairings automatically?

File: definitions/tools/pvacfuse.wdl
Comment (susannasiebert) [reply definitions/tools/pvacfuse.wdl:30]: I just released pVACtools 3.1.0 that has the combinatorial class II alleles enabled.

**Comment by susannasiebert** [comment]
3.1.0 is out now and has the required feature.

**Review by malachig — APPROVED** [review]
Looks good to me.

**Pull request merged**
